test-data/unit/check-based-misc.test

```diff
@@ -21,3 +21,46 @@ if True:
     1  # type: ignore[operator]
 [out]
 main:3:5:3:31: error: Unused "type: ignore" comment  [unused-ignore]
+
+
+[case testNarrowOnInitialAssignment]
+a: object = 1
+reveal_type(a)  # N: Revealed type is "int" (narrowed from "object")
+
+
+[case testNarrowWithAny]
+# mypy: allow-any-explicit, allow-any-expr
+from typing import Any, Union
+
+a: Union[int, Any]
+if bool():
+    a = 1
+    reveal_type(a)  # N: Revealed type is "int" (narrowed from "int | Any")
+b: Any = 1
+reveal_type(b)  # N: Revealed type is "Any"
+c: Any
+c = 1
+reveal_type(c)  # N: Revealed type is "Any"
+
+
+[case testNarrowWithTuple]
+# mypy: disallow-redefinition
+from typing import Tuple, TypeVar
+T = TypeVar('T')
+
+def f(x: T) -> Tuple[T, T]: ...
+x = None
+reveal_type(x) # N: Revealed type is "None" (narrowed from "partially defined: ? | None")
```

Comment:
`"partially defined: ? | None"` looks confusing, what exactly does it mean

Reply:
it means that the type isn't fully defined yet. It's currently inferred as something like `TheType[T] = T | None`, where the `T` hasn't been filled yet. Any ideas on how it could be improved?

Reply:
erm, after extensive back and forwarding, this is probably good as is.